### PR TITLE
feature: Partition support for DJLModel using SM Training job

### PR DESCRIPTION
**Pull request by sindhuvahinis**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: doc/frameworks/djl/using_djl.rst [hunk 1]
```diff
@@ -78,6 +78,24 @@ Inference data are serialized and sent to the DJL Serving model server by an ``I
...
 By default, the inference data is serialized to a json string, and the inference result is a Python dictionary.
 
+
+Before deploying the model, DJLModel enables you to partition the model in your desired instance. For larger models,
+runtime partitioning in DeepSpeed could increase the model loading time of models and CPU utilization significantly. So,
+for larger models, you could partition the models ahead of time. `partition` method makes use of Amazon Sagemaker
+Training job to invoke the partition of the model and upload it to S3 bucket. You can either provide your s3 bucket to
+upload checkpoints or it will be uploaded to default sagemaker S3 bucket. Please note that this s3 bucket will be
+remembered. When you call `deploy` method after partition, DJLServing downloads the model checkpoints directly from the
+uploaded s3 url.
```
[the remark this answers is not in the document]
Comment (sindhuvahinis) [reply doc/frameworks/djl/using_djl.rst:88]: Yeah I have been debating it myself. Sure, will ping you offline

[the remark this answers is not in the document]
Comment (sindhuvahinis) [reply doc/frameworks/djl/using_djl.rst:88]: Updated the document.

**Review by lanking520 — APPROVED** [review]

**Review by claytonparnell — COMMENTED** [review]
/bot run all

**Review by claytonparnell — COMMENTED** [review]
/bot run all

File: src/sagemaker/djl_inference/model.py [hunk 2]
```diff
@@ -398,6 +448,95 @@ def right_size(self, **_):
...
+            output_kms_key=output_kms_key,
+            use_spot_instances=use_spot_instances,
+            max_wait=max_wait,
+            enable_network_isolation=enable_network_isolation,
+        )
+
+        # creates a training job to do partitions
+        estimator.fit(
+            inputs=container_def["ModelDataUrl"],
+            wait=True,
```
Comment (siddvenk) [inline src/sagemaker/djl_inference/model.py:530]: Should we let the user configure this? I think by default we can make it blocking, but user may want to have this be async

File: src/sagemaker/djl_inference/model.py [hunk 3]
```diff
@@ -716,6 +872,8 @@ def __init__(
         self.enable_cuda_graph = enable_cuda_graph
         self.triangular_masking = triangular_masking
         self.return_tuple = return_tuple
+        self.save_mp_checkpoint_path = None
+        self.checkpoint = None
```
Comment (siddvenk) [inline src/sagemaker/djl_inference/model.py:876]: If user has a partitioned model that they didn't partition with sagemaker, they may want to specify this, right? I don't think we need to figure this out in this PR, but right now I don't think we support that use-case (maybe i'm wrong).

File: src/sagemaker/djl_inference/model.py [hunk 4]
```diff
@@ -924,3 +1196,60 @@ def __init__(
...
+        super(FasterTransformerModel, self).partition(
+            instance_type,
+            s3_output_uri,
+            job_name,
+            volume_kms_key=volume_kms_key,
+            output_kms_key=output_kms_key,
+            use_spot_instances=use_spot_instances,
+            max_wait=max_wait,
+            enable_network_isolation=enable_network_isolation,
+        )
```
Comment (siddvenk) [inline src/sagemaker/djl_inference/model.py:1255]: We don't need this, I don't see any FT specific logic. Parent implementation should be enough.

Comment (sindhuvahinis) [reply src/sagemaker/djl_inference/model.py:1255]: Okay

File: src/sagemaker/djl_inference/model.py (hunk 3 above)
Comment (sindhuvahinis) [reply src/sagemaker/djl_inference/model.py:876]: No you are right. In that case, we need to provide a way using PySDK. They could still work it out with their own serving.properties for now.

**Review by claytonparnell — COMMENTED** [review]
/bot run slow-tests, unit-tests

File: src/sagemaker/djl_inference/model.py (hunk 3 above)
Comment (sindhuvahinis) [reply src/sagemaker/djl_inference/model.py:876]: Resolved

File: src/sagemaker/djl_inference/model.py (hunk 4 above)
Comment (sindhuvahinis) [reply src/sagemaker/djl_inference/model.py]: Resolved

**Review by siddvenk — APPROVED** [review]

**Review by claytonparnell — COMMENTED** [review]
/bot run pr

**Review by knikure — COMMENTED** [review]
/bot run slow-tests, pr

**Review by knikure — APPROVED** [review]
/bot run pr

**Pull request merged**